### DISCUSSION TITLE
What are the r and rInv values for?

### DISCUSSION THREAD

**Discussion by nicolai-huening** [carbynestack/carbynestack#105] [Q&A] [answered]
Hello together, 
I have a question regarding the 'r' and 'rInv' values. I do not understand what part thees values play in the overall protocol. First I thought it is used for a kind of FHE-Offline-phase (As described [here](https://mp-spdz.readthedocs.io/_/downloads/en/latest/pdf/#HOMOMORPHIC%20ENCRYPTION)), but neither 'r' nor 'rInv' are used by klyshko, only by ephemeral and amphora.
As far as I can tell the r and rInv are inverse mod p, and they are used only on the shares of secrets. Is this done to ensure security during the communication between amphora and the client?

Thanks a lot in advance
Nicolai

**Comment by sbckr** [comment, marked as answer]
Hi @nicolai-huening.

`r` and `rInv` are used for conversion between BigInteger and _Montgomery Representation_  as used by MP-SPDZ to speed up computation. 
See [mp-spdz-integration](https://github.com/carbynestack/mp-spdz-integration/blob/master/src/main/java/io/carbynestack/mpspdz/integration/MpSpdzIntegrationUtils.java) for more details on how they are used internally.

Best regards
Sebastian

**Comment by sbckr** [comment]
@nicolai-huening 

By the way, thank you and congratulations for the first question in our organization's discussion space. 
# 🥳🪅🍾

**Reply by nicolai-huening** [reply]
Thank you very much, that was the puzzle pease I was missing! 😄
